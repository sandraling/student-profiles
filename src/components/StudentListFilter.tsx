import * as React from "react";
import FormControl from '@material-ui/core/FormControl';
import Input from '@material-ui/core/Input';

interface StudentListFilterProps {
    setNameFilter: (text: string) => void;
    setTagFilter: (text: string) => void;
}

export const StudentListFilter = (props: StudentListFilterProps) => {
    const { setNameFilter, setTagFilter } = props;

    return (
        <form className="student-profile__search-form">
            <Input 
                id="name-input" 
                className="student-profile__search-bar"
                placeholder="Search by name"
                onChange={(e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => setNameFilter(e.target.value)}
                />
            <Input 
                id="tag-input" 
                className="student-profile__search-bar"
                placeholder="Search by tag"
                onChange={(e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => setTagFilter(e.target.value)}
            />
        </form>
    )
}