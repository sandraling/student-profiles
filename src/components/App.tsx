import * as React from "react";
import { BrowserRouter, Route, Switch } from 'react-router-dom';

import { createMuiTheme, ThemeProvider, StylesProvider } from '@material-ui/core/styles';
import { grey } from '@material-ui/core/colors';

import '../index.css';
import { StudentProfiles } from './StudentProfiles';

export const App = () => {
    // Change material ui theme to grey
    const theme = createMuiTheme({
        palette: {
          primary: {
            main: grey[700],
          },
        },
    });

    return (
    <>
        <StylesProvider injectFirst> {/* Ensure css takes precedance over mui styling */}
            <ThemeProvider theme={theme}>
                <BrowserRouter>
                <Switch>
                    <Route exact path="/" component={StudentProfiles} />
                    <Route render={() => <h1>404</h1>} />
                </Switch>
                </BrowserRouter>
            </ThemeProvider>
        </StylesProvider>
    </>
    );
};