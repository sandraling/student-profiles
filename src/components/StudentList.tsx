import * as React from "react";
import List from '@material-ui/core/List';
import { CircularProgress } from "@material-ui/core";
import { StudentListItem } from './StudentListItem';

export interface StudentObj {
    city: string;
    company: string;
    email: string;
    firstName: string;
    grades: number[];
    id: number;
    lastName: string;
    pic: string;
    skill: string;
    tags: string[];
    expanded: boolean;
}

interface StudentListItemProps {
    students: StudentObj[];
    hiddenStudents: number[];
    expandOnClick: (id: number) => void;
    addTagToStudent: (id: number, tag: string) => void;
}

export const StudentList = (props: StudentListItemProps) => {
    const { students, hiddenStudents, expandOnClick, addTagToStudent } = props;
    let count = 0;

    if (students.length === 0) {
        return (
            <CircularProgress className="student-profile__list--loading" size={60} />
        );
    }

    return (
        <List
            aria-labelledby="List of students"
            className="student-profile__list-container"
        >
            {
                students.map((student) => {
                    let studentListItem = null;
                    if (hiddenStudents.includes(student.id) === false) {
                        studentListItem = (
                            <StudentListItem 
                                student={student}
                                index={count}
                                expandOnClick={(id) => expandOnClick(id)} 
                                addTagToStudent={(id, tag) => addTagToStudent(id, tag)}
                                key={`student-list-item-${student.id}`}
                            />
                        )
                        count += 1;
                    }
                    return studentListItem;                    
                })
            }
        </List>        
    )
}