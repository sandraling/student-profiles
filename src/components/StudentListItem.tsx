import * as React from "react";
import ListItem from '@material-ui/core/ListItem';
import ListItemAvatar from '@material-ui/core/ListItemAvatar';
import Divider from '@material-ui/core/Divider';
import Avatar from '@material-ui/core/Avatar';
import Typography from '@material-ui/core/Typography';
import Collapse from '@material-ui/core/Collapse';
import IconButton from '@material-ui/core/IconButton';
import Chip from '@material-ui/core/Chip';
import Input from '@material-ui/core/Input';
import { faPlus, faMinus } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { StudentObj } from './StudentList';

interface StudentListItemProps {
    student: StudentObj;
    index: number;
    expandOnClick: (id: number) => void;
    addTagToStudent: (id: number, tag: string) => void;
}

/**
 * Component representing one single student's profile
 * @param props see StudentListItemProps interface
 */
export const StudentListItem = (props: StudentListItemProps) => {
    const { student, index, expandOnClick,addTagToStudent } = props;
    const fullName = student.firstName + " " + student.lastName;
    const [tagInput, setTagInput] = React.useState<string>("");
    const onFormSubmit = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        // Some error checking before saving tag to student.
        // Error not alternatively presented since this is 
        // out of the scope of this project!
        if (tagInput !== "" && !student.tags.includes(tagInput)) {
            addTagToStudent(student.id, tagInput);
            setTagInput("");
        }
    }

    return (
        <>  
            { index === 0 ? null : <Divider /> }
            <ListItem className="student-profile__list-item" key={`student-${student.id}`}>
                <ListItemAvatar>                        
                    <Avatar 
                        className="student-profile__avatar"
                        src={student.pic} 
                        alt={`Profile picture of ${fullName}`} 
                    />
                </ListItemAvatar>
                <Typography component="div" variant="body2" className="student-profile__main-body">
                    <div className="student-profile__name">{fullName}</div>
                        <>
                            <span className="student-profile__description">
                                <span>Email: {student.email}</span>
                                <span>Company: {student.company}</span>
                                <span>Skill: {student.skill}</span>
                                <span>Average: {React.useMemo(() => getAverage(student.grades), [student.id])}%</span>
                                <Collapse in={student.expanded} timeout="auto" unmountOnExit>
                                    <div className="student-profile__subsection">
                                        <table className="student-profile__subsection__grades-table" key={`student-${student.id}-grade-table`}>
                                            <tbody>
                                                {student.grades.map((grade, index) => (
                                                    <tr key={`student-${student.id}-grade-tr-${index}`}>
                                                        <td>{`Test ${index}:`}</td>
                                                        <td>{`${grade}%`}</td>
                                                    </tr>
                                                ))}
                                            </tbody>
                                        </table>
                                        <div className="student-profile__subsection__tag-div">
                                            {
                                                student.tags.map((tag, index) => 
                                                    <Chip 
                                                        key={`student-${student.id}-chip-${index}`}
                                                        size="small" 
                                                        label={tag}
                                                        className="student-profile__subsection__tag-btn"
                                                    />
                                                )
                                            }
                                        </div>
                                        <form onSubmit={(e) => onFormSubmit(e)}>
                                            <Input 
                                                className="add-tag-input student-profile__subsection__tag-input"
                                                placeholder="Add a tag"
                                                value={tagInput}
                                                onChange={(e) => setTagInput(e.target.value)}
                                            />
                                        </form>
                                    </div>
                                </Collapse>
                            </span>
                        </>
                </Typography>
                <IconButton 
                    aria-label="expand" 
                    className="expand-btn student-profile__expand-btn"
                    onClick={() => expandOnClick(props.student.id)}
                >
                    {student.expanded
                        ? <FontAwesomeIcon icon={faMinus}/> 
                        : <FontAwesomeIcon icon={faPlus} />}
                </IconButton>
            </ListItem>    
        </>
    )
}

/**
 * Given a list of numbers, return the average of them.
 * @param numList An array of numbers
 */
export const getAverage = (numList: number[]) => {
    const sum = numList.reduce((a, b) => a + b, 0);
    return (sum / numList.length) || 0;
}