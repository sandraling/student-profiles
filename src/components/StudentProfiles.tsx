import * as React from "react";
import axios from "axios";
import * as _ from "lodash";
import { CircularProgress} from "@material-ui/core";
import { StudentListFilter } from './StudentListFilter';
import { StudentObj, StudentList } from './StudentList';

const studentDataSource = 'https://www.hatchways.io/api/assessment/students';

interface StudentData {
    students: StudentDataObj[];
}

interface StudentDataObj {
    city: string;
    company: string;
    email: string;
    firstName: string;
    grades: string[];
    id: string;
    lastName: string;
    pic: string;
    skill: string;
    tag: string[];
    expanded: boolean;
}

/**
 * Outputs the student profile page
 */
export const StudentProfiles = () => {
    const [loading, setLoading] = React.useState<boolean>(false);
    const [students, setStudents] = React.useState<StudentObj[]>([]);
    const [error, setError] = React.useState<boolean>(false);
    const [errorMsg, setErrorMsg] = React.useState<string>('');
    const [nameFilter, setNameFilter] = React.useState<string>('');
    const [tagFilter, setTagFilter] = React.useState<string>('');
    const [editRevision, setEditRevision] = React.useState<number>(0);
    // Each time tags are modified, tag revision goes up by 1
    
    /**
     * Fetching student data from server
     */
    React.useEffect(() => {
        axios.get<StudentData>(studentDataSource)
            .then(({ data }) => {
                const students = data.students.map((student) => ({
                    ...student,
                    id: parseInt(student.id),
                    grades: student.grades.map(grade => parseInt(grade)),
                    tags: [],
                    expanded: false
                }));
                setLoading(false);
                setStudents(students);
            })
            .catch(() => {
                setError(true);
                setErrorMsg("We couldn't find the student profiles this time.")
                setLoading(false);
            })
    }, [])

    /**
     * Given id of student and tag, add tag to student if student with
     * said id exists. Update error and error message hooks otherwise.
     * @param id student id
     * @param tag tag to be added to student
     */
    const addTagToStudent = (id: number, tag: string) => {
        const targetIndex = students.findIndex(student => student.id === id);
        if (targetIndex === -1) {
            setError(true);
            setErrorMsg("Failed to set add a tag to a student.");
        } else {
            setStudents(students => {
                const target = students[targetIndex]
                students[targetIndex] = {
                    ...target,
                    tags: [...target.tags, tag]
                }
                return students;
            })
            setEditRevision(rev => rev + 1);
        }
    }

    /**
     * An array of student id's filtered out by name
     */
    const filterByName = React.useMemo(() => {
        let results: number[] = [];
        students.forEach(student => {
            const fullName = (student.firstName + " " + student.lastName).toLowerCase();
            if (fullName.includes(nameFilter) === false) results.push(student.id);
        })
        return results;
    }, [nameFilter])

    /**
     * An array of student id's filtered out by tag
     */
    const filterByTag = React.useMemo(() => {
        let results: number[] = [];
        students.forEach(student => {
            const relevantTags = student.tags.filter(tag => tag.includes(tagFilter));
            if (relevantTags.length === 0) results.push(student.id);
        })
        return results;
    }, [tagFilter, editRevision])

    /**
     * An array of student id's filtered out by both name and tag;
     * Students not filtered out (i.e. not in the array) will either match
     * by name or tag or both.
     */
    const filterResults = () => {
        if (tagFilter === "") return filterByName;
        if (nameFilter === "") return filterByTag;
        return _.intersection(filterByName, filterByTag);
    }

    /**
     * Hook call functions passed down to child components
     */
    const expandProfile = (id: number) => {
        setStudents(students => 
            students.map(student => 
                student.id === id
                    ? {
                        ...student,
                        expanded: !student.expanded
                    }
                    : student))
    }

    /**
     * Student profile rendering
     */    
    if (loading) {
        return (
            <div>
                <CircularProgress />
            </div>
        )
    }

    if (error === true) {
        return (
            <div>
                <h1>Oh no!</h1>
                <h2>{errorMsg}</h2>
            </div>
        )
    }

    return (
        <div className="student-profile__container">
            <StudentListFilter 
                setNameFilter={(text) => setNameFilter(text)}
                setTagFilter={(text) => setTagFilter(text)}
            />
            <StudentList 
                students={students}
                hiddenStudents={filterResults()}
                expandOnClick={(id) => expandProfile(id)}
                addTagToStudent={(id, tag) => addTagToStudent(id, tag)}
            />
        </div>
    )
}