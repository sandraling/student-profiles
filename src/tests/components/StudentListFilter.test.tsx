import * as React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import { StudentListFilter } from "../../components/StudentListFilter";

describe("<StudentListFilter />", () => {
    let wrapper: ShallowWrapper;
    const setNameFilter = jest.fn((text: string) => {});
    const setTagFilter = jest.fn((text: string) => {});

    beforeEach(() => {
        wrapper = shallow(
            <StudentListFilter 
                setNameFilter={setNameFilter}
                setTagFilter={setTagFilter}
            />
        )
    });
    
    it("renders default state successfully", () => {
        expect(wrapper).toMatchSnapshot();
    })

    it("calls setNameFilter whenever name input field is changed", () => {
        const newName = "new name";
        wrapper.find("#name-input").at(0).simulate("change", {
            target: { value: "new name" }
        });
        expect(setNameFilter).toHaveBeenCalledWith(newName);
    })

    it("calls setTagFilter whenever tag input field is changed", () => {
        const newTag = "new tag";
        wrapper.find("#tag-input").at(0).simulate("change", {
            target: { value: "new tag" }
        });
        expect(setTagFilter).toHaveBeenCalledWith(newTag);
    })

});