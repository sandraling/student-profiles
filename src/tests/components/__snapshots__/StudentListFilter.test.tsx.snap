// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StudentListFilter /> renders default state successfully 1`] = `
<form
  className="student-profile__search-form"
>
  <WithStyles(ForwardRef(Input))
    className="student-profile__search-bar"
    id="name-input"
    onChange={[Function]}
    placeholder="Search by name"
  />
  <WithStyles(ForwardRef(Input))
    className="student-profile__search-bar"
    id="tag-input"
    onChange={[Function]}
    placeholder="Search by tag"
  />
</form>
`;
